# Running SwarmGenomics
SwarmGenomics offers two ways to process whole-genome data, depending on your workflow preference and time constraints:

**1. Step-by-Step Execution** – Run each module and step separately, allowing for greater flexibility in selecting specific analyses. This is ideal for customized workflows or focusing on particular aspects of the genome.

**2. Fast-Track Execution** – Use pre-prepared scripts to automate multiple steps in a streamlined process. This option is designed to be easier and more efficient.

Choose the method that best suits your research needs, whether you prefer full control over each step or a faster, automated approach. Below, you'll find detailed instructions and scripts for the fast-track execution. Otherwise go to step **01. Getting started** for slower approach.

## Preprocessing
By now you should already have all the software installed as instructed in **0. Installations**.
Before starting the preprocessing script, you should download the reference genome in FASTA format and the FASTQ files

### Downloading data
This is an example with golden eagle genome. Change the details according to your species.
You can use a text editor, such as Notepad, to edit the script and save it. 

#### Create a directory for your species in your working directory
When running SwarmGenomics, it's important to be in the right folder (directory) — this is called your working directory. It's the place where the commands will create files, download data, and look for input files. If you're not in the correct directory, things may not work as expected, or files might end up in the wrong place.

In the SwarmGenomics course the working directory will be /vol/storage/swarmgenomics/your_name/

```
mkdir golden_eagle
```

#### Change to that directory
```
cd golden_eagle
```

#### Download the reference genome from https://www.ncbi.nlm.nih.gov/genome/
When you find your species, go to FTP and copy the link address of the file that ends with *.fna.gz*
```
wget -O ./reference_genome.fna.gz address_to_.fna.gz file
wget -O /vol/storage/swarmGenomics/golden_eagle/ac_reference.fna.gz https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/900/496/995/GCF_900496995.4_bAquChr1.4/GCF_900496995.4_bAquChr1.4_genomic.fna.gz
```
#### Choose and download the SRA file from with accession number https://www.ncbi.nlm.nih.gov/sra
Copy the run ID
```
prefetch ERR3316068 
```
If you get an error due to size
```
prefetch --verbose --force all accession_number --max-size 50G
```
### Run preprocessing.bash
Download the preprocessing.bash script and edit the paths to: WORKING_DIR, ADAPTERS, BCFTOOLS, and SRATOOLS. Also adjust the number of threads according to your available resources.
```
# Check how many CPU cores (threads) available
nproc
```
Once you have edited the the script, you can upload it to your working directory. 
```
# Make the file executable
chmod +x preprocessing.bash

# If the script is written or edited on Windows and then run on a Unix system, run:
dos2unix preprocessing.bash
```
The scrip takes around 24h to run depending on available resources, so you should use nohup and & to keep the it running in the background without stopping if the terminal is closed.
```
nohup ./preprocessing.bash "species" "reference_genome.fna.gz" "SRA file ID" &

# E.g with golden eagle
nohup ./preprocessing.bash "golden_eagle" "/vol/storage/swarmgenomics/reference.fna.gz" "ERR3316068" &
```
To check on the progress use htop and read the nohup.out file in case of errors.
```
# Check progress and if the script is running
htop

# Read the nohup.out 
less nohup.out
```
## Genomic modules
Once the preprocessing is done you can continue with the genomic modules, so first steps 3.-6: Genome Features, Heterozygosity, Runs of Homozygosity and PSMC.

### Installations
Download genomic_modules.bash, all_roh_bar_plots.R, idxstats_summary_plot.R and make sure you have all necessary installations completeted:
bcftools, R packages and PSMC.
```
# Install R
conda install r-essentials r-base

# Open R
R

# Install packages
# Choose CRAN mirror, e.g. 39
install.packages(c("ggplot2", "dplyr", "gridExtra", "data.table","RColorBrewer", "readr", "grid", "gridExtra"))
```
Download PSMC in your software directory (e.g. /vol/storage/software)
```
# Clone the repository
git clone https://github.com/lh3/psmc.git

# Make psmc
cd psmc
make

# Make utils
cd utils
make
```
PSMC plot for in your psmc directory (e.g. /vol/storage/software/psmc)
```
# Clone the repository
git clone https://github.com/willyrv/ms-PSMC.git

# Change permissions
chmod +x ./*
```
### Editing genomic_modules.bash
Make sure genomic_modules.bash and the plotting scripts (all_roh_bar_plots.R and idxstats_summary_plot.R) are in your species directory and edit the paths according to your directories.

In **genomic_modules.bash** edit the marked paths:
```
#!/bin/bash
# Version 1.0 
# 21.05.2025

# Input variables
SPECIES=$1

# Directories
WORKING_DIR="/vol/storage/swarmgenomics" # Change this
VCF_DIR="${WORKING_DIR}/${SPECIES}/vcf"
VCF_HET_DIR="${WORKING_DIR}/${SPECIES}/vcf_het"
ROH_DIR="${WORKING_DIR}/${SPECIES}/roh"
RESULTS_DIR="${WORKING_DIR}/${SPECIES}/results"

# Tools
BCFTOOLS="/vol/storage/software/bcftools-1.17/bcftools" # Change this
PSMC="/vol/storage/software/psmc/psmc" # Change this
PSMC_UTILS="/vol/storage/software/psmc/utils" # Change this

# File paths
BAM_FILE="${WORKING_DIR}/${SPECIES}/bwa.sorted.bam" # Change this if different bam file name
```
Once scripts have been edited and uploaded, make them executable and in correct format. 
```
# Change permissions
chmod +x genomic_modules.bash all_roh_bar_plots.R idxstats_summary_plot.R

# If you edited script on notepad etc change format
dos2unix genomic_modules.bash all_roh_bar_plots.R idxstats_summary_plot.R
```
### Running genomic_modules.bash
To run the script 
```
nohup ./genomic_modules.bash "your_species" &

# For example
nohup ./genomic_modules.bash "giant_panda" &
```
### Output
The script combines modules Genome Features, Heterozygosity, Runs of Homozygosity and PSMC. The outputs are a PNG file with plots from idxstats results and heterozygosity, a PNG with bar plots fo the RoH results, and a PDF file with PSMC results. 

PSMC is unscaled by default, but to scale it according to your species generation time and mutation rate, add the infromation on this line of the genomic_modules.bash:
```
# Unedited version
"${PSMC_UTILS}/psmc_plot.pl" -p "${WORKING_DIR}/${SPECIES}/diploid" "${DIPLOID_PSMC}"

# With mutation rate and generation time
"${PSMC_UTILS}/psmc_plot.pl" -u 1.29e-8 -g 12 -p "${WORKING_DIR}/${SPECIES}/diploid" "${DIPLOID_PSMC}"
```
## Unmapped Reads
## Mitogenome assembly and NUMTs
## Repeat Analysis
