# Heterozygosity
Understanding genetic diversity within a population is crucial for addressing the complexities of evolution, adaptation, and conservation. Heterozygosity, one of the most significant indicators of genetic diversity, measures the presence of different alleles at a gene locus within an individualâ€™s genome. High heterozygosity indicates healthy genetic variation, allowing populations to adapt to environmental changes and avoid inbreeding depression. While, low heterozygosity may suggest inbreeding, genetic drift, or population bottlenecks, which can increase the risk of extinction.

## Installations
You should already have bcftools installed, otherwise you will only need to download and edit calculate_heterozygosity.sh and plot_heterozygosity.py.

In calculate_heterozygosity.sh, edit:
```
# Path to working directory
VCF_DIR="/vol/storage/swarmGenomics/giant_panda/vcf_het"

# And the path to bcftools
/vol/storage/bcftools-1.19/bcftools
```
## Running the scripts
```
# Make a directory for the vcf files
mkdir vcf_het

# Copy the vcf files from /vcf
cd vcf
cp *.vcf /vol/storage/swarmGenomics/golden_eagle/vcf_het/

# Change back to working directory
# You should have the two edited scripts downloaded there
# Change permissions
chmod +x calculate_heterozygosity.sh

# Run calculate_heterozygosity.sh
./calculate_heterozygosity.sh

# Once finished you can check the output "heterozygosity_results.txt", it should have the scaffold/chromosome name and corresponding heterozygosity
# Then plot the results
python plot_heterozygosity.py
```

### Output
The output graph shows heterozygosity per largest scaffolds/chromosomes.
