# Heterozygosity
Understanding genetic diversity within a population is crucial for addressing the complexities of evolution, adaptation, and conservation. Heterozygosity, one of the most significant indicators of genetic diversity, measures the presence of different alleles at a gene locus within an individualâ€™s genome. High heterozygosity indicates healthy genetic variation, allowing populations to adapt to environmental changes and avoid inbreeding depression. While, low heterozygosity may suggest inbreeding, genetic drift, or population bottlenecks, which can increase the risk of extinction.

## Installations
You should already have bcftools installed, otherwise you will only need to download and edit **calculate_heterozygosity.sh** and **plot_heterozygosity.R**.

In **calculate_heterozygosity.sh**, edit:
```
# Path to working directory
VCF_DIR="/vol/storage/swarmGenomics/giant_panda/vcf" # change this
OUTPUT_FILE="/vol/storage/swarmGenomics/giant_panda/heterozygosity_results.txt" # change this
BCFTOOLS="/vol/storage/software/bcftools-1.19" # change this
```
## Running the scripts on multiple VCF files (SwarmGenomics preprocessing pipeline)
The script runs through the vcf files created in the preprocessing steps of the SwarmGenomics pipeline.
```
# You should have the two edited scripts downloaded there
# Change permissions
chmod +x calculate_heterozygosity.sh

# Run calculate_heterozygosity.sh
nohup ./calculate_heterozygosity.sh &

# Once finished you can check the output "heterozygosity_results.txt", it should have the scaffold/chromosome name and corresponding heterozygosity
# Then plot the results
Rscript plot_heterozygosity.R
```
### Output
The output graph shows heterozygosity per largest scaffolds/chromosomes.

## Estimating Heterozygosity on a single VCF file
If you want to quickly calculate heterozygosity for a single VCF file without running the full script, you can use the following one-liner:
```
# Give a path to bcftools if needed
# Change file.vcf according to your file name
bcftools query -f '[%GT\n]' file.vcf | awk '{het+=($0=="0/1"||$0=="1/0"); hom+=($0=="0/0"||$0=="1/1")} END{if(het+hom>0) print het/(het+hom); else print 0}'
```
This will simply then print the calculated heterozygosity value.
