# Preparing directories and downloading data 

## Objectives
1. Create directories for storing your data 
2. Download the reference genome of your choice 
3. Download the SRA file for your species
4. Prepare the data for mapping

### Downloading data
This is an example with Golden eagle genome. Change the details according to your species.
You can use a text editor, such as Notepad, to edit the script and save it. 

#### Create a directory for your species (your working directory)
```
mkdir golden_eagle
```
#### Download the reference genome from https://www.ncbi.nlm.nih.gov/genome/
```
wget -O /vol/storage/swarmGenomics/golden_eagle/ac_reference.fna.gz https://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/900/496/995/GCF_900496995.4_bAquChr1.4/GCF_900496995.4_bAquChr1.4_genomic.fna.gz
```
#### Choose and download the SRA file from with accession number https://www.ncbi.nlm.nih.gov/sra
```
prefetch ERR3316068 
```
If you get an error due to size
```
prefetch --verbose --force all accession_number --max-size 50G 
```
### Preparing data for mapping
#### Converting into fastq
```
fastq-dump --outdir /vol/storage/swarmGenomics/golden_eagle/fastq --split-files /vol/storage/swarmGenomics/golden_eagle/ERR3316068/ERR3316068.sra
```
#### Quality control
```
fastqc -t 13 -o /vol/storage/swarmGenomics/golden_eagle  -f fastq /vol/storage/swarmGenomics/golden_eagle/fastq/ERR3316068_1.fastq /vol/storage/swarmGenomics/golden_eagle/fastq/ERR3316068_2.fastq
```
#### Check output
First copy it to your local storage
```
scp -r -P 30189 -i '/path/to/your/private_key.txt' ubuntu@134.176.27.78:/vol/storage/swarmGenomics/golden_eagle/ERR3316068_1_fastqc* .
```
Then open the html file and have a look at the output
You can look up information on fastqc to see how to interpret the results

#### Trimming

Trimmomatic removes low quality reads and adapter sequences. 
```
trimmomatic PE -threads 13 /vol/storage/swarmGenomics/golden_eagle/fastq/ERR3316068_1.fastq /vol/storage/swarmGenomics/golden_eagle/fastq/ERR3316068_2.fastq /vol/storage/swarmGenomics/golden_eagle/fastq/ERR3316068_1_paired.fastq.gz /vol/storage/swarmGenomics/golden_eagle/fastq/ERR3316068_1_unpaired.fastq.gz /vol/storage/swarmGenomics/golden_eagle/fastq/ERR3316068_2_paired.fastq.gz /vol/storage/swarmGenomics/golden_eagle/fastq/ERR3316068_2_unpaired.fastq.gz SLIDINGWINDOW:4:20 MINLEN:25 ILLUMINACLIP:/vol/storage/software/trimmomatics/adapters.fa:2:30:10:2:keepBothReads
```
Have a look at http://www.usadellab.org/cms/index.php?page=trimmomatic if you're curious about the different steps.

#### Second quality control

Adapt the script from first quality control to run it with the output files from trimmomatics, and then move the files to your local storage for checking as you did before.

#### Create a fastqc directory and move your fastqc files there
Commands you need:
```
mkdir
mv
```
