# Mapping the reference genome

#### Map reads onto reference
```
bwa index /vol/storage/swarmGenomics/golden_eagle/acreference.fna.gz
```

#### Convert into a bam file
```
#Align paired-end reads to reference genome
bwa mem -t 13 /vol/storage/swarmGenomics/golden_eagle/acreference.fna.gz /vol/storage/swarmGenomics/golden_eagle/fastq/ERR3316068_1_paired.fastq.gz /vol/storage/swarmGenomics/golden_eagle/fastq/ERR3316068_2_paired.fastq.gz > /vol/storage/swarmGenomics/golden_eagle/ERR3316068.sam

#Sort aligned reads and save as BAM
samtools sort -@ 13 /vol/storage/swarmGenomics/golden_eagle/ERR3316068.sam -o /vol/storage/swarmGenomics/golden_eagle/bwa.sorted.bam

#Index sorted BAM file
samtools index /vol/storage/swarmGenomics/golden_eagle/bwa.sorted.bam
```
#### Index the reference
```
#Unzip the reference for samtools
gunzip /vol/storage/swarmGenomics/golden_eagle/acreference.fna.gz

#Index the reference
samtools faidx /vol/storage/swarmGenomics/golden_eagle/acreference.fna
```
