# Mapping reads onto the reference genome

#### Map reads onto reference
```
bwa index /vol/storage/swarmGenomics/golden_eagle/acreference.fna.gz
```

#### Convert into a bam file
```
#Align paired-end reads to reference genome
bwa mem -t 13 /vol/storage/swarmGenomics/golden_eagle/acreference.fna.gz /vol/storage/swarmGenomics/golden_eagle/fastq/ERR3316068_1_paired.fastq.gz /vol/storage/swarmGenomics/golden_eagle/fastq/ERR3316068_2_paired.fastq.gz > /vol/storage/swarmGenomics/golden_eagle/ERR3316068.sam

#Sort aligned reads and save as BAM
samtools sort -@ 13 /vol/storage/swarmGenomics/golden_eagle/ERR3316068.sam -o /vol/storage/swarmGenomics/golden_eagle/bwa.sorted.bam

#Index sorted BAM file
samtools index /vol/storage/swarmGenomics/golden_eagle/bwa.sorted.bam
```
#### Index the reference
```
#Unzip the reference for samtools
gunzip /vol/storage/swarmGenomics/golden_eagle/acreference.fna.gz

#Index the reference
samtools faidx /vol/storage/swarmGenomics/golden_eagle/acreference.fna
```
#### Variant calling
```
/vol/storage/bcftools-1.19/bcftools mpileup -C50 -f /vol/storage/swarmGenomics/golden_eagle/acreference.fna  /vol/storage/swarmGenomics/golden_eagle/bwa.sorted.bam | /vol/storage/bcftools-1.19/bcftools call -c -o /vol/storage/swarmGenomics/golden_eagle/output.vcf 
```

#### Remove repeats
```
#Create bed file with repeats
perl -lne 'if(/^(>.*)/){ $head=$1 } else { $fa{$head} .= $_ } END{ foreach $s (sort(keys(%fa))){ print "$s\n$fa{$s}\n" }}'  /vol/storage/swarmGenomics/golden_eagle/acreference.fna | perl -lne 'if(/^>(\S+)/){ $n=$1} else { while(/([a-z]+)/g){ printf("%s\t%d\t%d\n",$n,pos($_)-length($1),pos($_)) } }'  > /vol/storage/swarmGenomics/golden_eagle/repeats.bed

#Remove repeats from vcf
bedtools subtract -header -a /vol/storage/swarmGenomics/golden_eagle/output.vcf -b /vol/storage/swarmGenomics/golden_eagle/repeats.bed > /vol/storage/swarmGenomics/golden_eagle/output_no_repeats.vcf

#make vcf without repeats standard vcf file
mv /vol/storage/swarmGenomics/golden_eagle/output_no_repeats.vcf /vol/storage/swarmGenomics/golden_eagle/output.vcf
```
