# Creating a file with largest chromosomes
To save time on the analyses we will be working on chromosomes/scaffold of >5MB.

#### Select for >5MB chromosomes
You may alter the code to select for chromosomes of different size.

```
perl -lne 'if(/^(>.*)/){ $head=$1 } else { $fa{$head} .= $_ } END{ foreach $s (sort(keys(%fa))){ print "$s\n$fa{$s}\n" }}'  /vol/storage/swarmGenomics/golden_eagle/acreference.fna | perl -lne 'if(/^>(\S+)/){ $n=$1} else { while(/([a-z]+)/g){ printf("%s\t%d\t%d\n",$n,pos($_)-length($1),pos($_)) } }'  > /vol/storage/swarmGenomics/golden_eagle/repeats.bed
```
#### Gzip and index the file
```
/vol/storage/bcftools-1.19/bcftools view /vol/storage/swarmGenomics/golden_eagle/output.vcf -Oz -o /vol/storage/swarmGenomics/golden_eagle/output.vcf.gz

#index vcf file (output.vcf.gz.tbi)
tabix -p vcf /vol/storage/swarmGenomics/golden_eagle/output.vcf.gz
```
#### VCF Filtering and Conversion to FASTQ
```
mkdir /vol/storage/swarmGenomics/golden_eagle/vcf
cd /vol/storage/swarmGenomics/golden_eagle/vcf

#Select larger chromosomes
while read chr; do tabix -h /vol/storage/swarmGenomics/golden_eagle/output.vcf.gz $chr > /vol/storage/swarmGenomics/golden_eagle/vcf/$chr.vcf ;done < /vol/storage/swarmGenomics/golden_eagle/target_len.lst

#Filtering
for file in *; do vcftools --vcf $file --minDP 5 --min-alleles 2 --max-alleles 2 --recode --recode-INFO-all --out ${file%.vcf}_filtered; done;

#vcf to fq
for file in *_filtered.recode.vcf; do bgzip -@ 10 $file; done;
for file in *_filtered.recode.vcf.gz; do /vol/storage/bcftools-1.19/bcftools index /vol/storage/swarmGenomics/golden_eagle/$file; done;
while read chr; do samtools faidx /vol/storage/swarmGenomics/golden_eagle/acreference.fna $chr | /vol/storage/bcftools-1.19/bcftools consensus ${chr}_filtered.recode.vcf.gz > ${chr}.fq ;done < /vol/storage/swarmGenomics/test/target_len.lst

cat *.fq > diploid.fq

#Gzip the file and move to main directory
gzip diploid.fq
mv /golden_eagle
```
